File: src/P052/p052.py
import itertools


def solve() -> int:
    for base in itertools.count(start=2):
        limit = 10 ** (base + 1)
        for num in itertools.count(start=10**base):
            if num % 3 != 0:
                continue
            if num * 6 >= limit:
                break
            digits = sorted(str(num))
            if "0" not in digits and "5" not in digits:
                continue
            if "2" not in digits and "3" not in digits:
                continue
            if "3" not in digits and "4" not in digits:
                continue
            if "4" not in digits and "5" not in digits:
                continue
            if "5" not in digits and "6" not in digits:
                continue
            if "6" not in digits and "7" not in digits:
                continue
            if all(sorted(str(num * i)) == digits for i in range(2, 7)):
                return int(num)
    raise RuntimeError("No solution found")

File: src/P052/p052.go
package problem052

import (
	"sort"
	"strconv"

	"github.com/sunilbpandey/project-euler/src/utils/go/math/intutils"
	"github.com/sunilbpandey/project-euler/src/utils/go/sliceutils"
)

func Solve() string {
	pow := 2
	for ; ; pow++ {
		limit := intutils.Pow(10, pow+1)
		start := intutils.Pow(10, pow)
		for n := start; 6*n < limit; n++ {
			if n%3 != 0 {
				continue
			}

			nDigits := intutils.Digits(n)
			digitCounts := make([]int, 10)
			for _, d := range nDigits {
				digitCounts[d]++
			}

			if (digitCounts[0] == 0 && digitCounts[5] == 0) ||
				(digitCounts[2] == 0 && digitCounts[3] == 0) ||
				(digitCounts[3] == 0 && digitCounts[4] == 0) ||
				(digitCounts[4] == 0 && digitCounts[5] == 0) ||
				(digitCounts[5] == 0 && digitCounts[6] == 0) ||
				(digitCounts[6] == 0 && digitCounts[7] == 0) {
				continue
			}

			sort.Ints(nDigits)
			sameDigits := true
			for i := 2; i <= 6 && sameDigits; i++ {
				mDigits := intutils.Digits(i * n)
				sort.Ints(mDigits)
				sameDigits = sliceutils.Equal(mDigits, nDigits)
			}
			if sameDigits {
				return strconv.Itoa(n)
			}
		}
	}
}
